### ISSUE TITLE
support adding/transfering data straight to cache/remote 

### ISSUE THREAD

**Issue by efiop** [iterative/dvc#4520]
We often see people trying to use `--external` to add some big dataset that they have on an external drive, where they also have an external cache dir. People often do that because they can't/don't want to copy their data to dvc repo to `dvc add` it normally, e.g. because their HDD/SSD won't be able to physically fit two copies of that dataset.

Same thing with s3/gs/etc, where people want to just move data straight to their remote, without having to download/add/push it, because, again, it might not even fit on their local machine.

That's why it would be great to introduce feature(s) to be able to move(or copy) data straight to cache/remote from it's original location. Potentially this is not only useful for `dvc add` but also for `dvc import[-url]`, where you want to use some data (e.g. through streaming with our API) in your project, that won't fit on your machine.

Related to https://github.com/iterative/dvc/issues/3920

**Comment by jorgeorpinel** [comment]
Would this also solve #3305?

**Comment by jorgeorpinel** [comment]
And yes, this feature would be useful. It sounds like a "lightweight" data registry (see https://dvc.org/doc/use-cases/data-registries) for people to use DVC as an intermediary between data already hosted on cloud storage and other projects. So they can access the data with a unified interface (DVC CLI or API).

**Comment by efiop** [comment]
> Would this also solve #3305?

No, it won't. These are unrelated. Unless I'm missing something.

**Comment by tomasfarias** [comment]
Hey all, I'll be tackling this issue next 👀

Let me know if I understand the problem clearly: although datasets in external drives/s3/etc can be `dvc add`ed with `--external` and later `dvc push`ed to a remote, we would like to streamline the process via a new feature(s). 

From the users perspective, I think something like a `dvc transfer` command that combines a download of the data, `dvc add` and `dvc push` would work well. Of course, under the hood we would not actually download the whole dataset, to avoid size issues as @efiop described.

An initial idea I had involved downloading chunks of a dataset at a time to avoid having to download the whole dataset. I'm aware at least s3 and WebHDFS support iterating over chunks as [`botocore`](https://github.com/boto/botocore) has an [`iter_chunks`](https://github.com/boto/botocore/blob/master/botocore/response.py#L104) method for it, and `HdfsCLI`'s `Client.read` supports a [`chunk_size`](https://github.com/mtth/hdfs/blob/master/hdfs/client.py#L642) argument. 

Essentially, a method `iter_chunks` that iterates over chunks of the file of size `chunk_size` could be added to each tree class in `dvc/tree/`, and a `transfer` method can then write the outputs of the generator as they come. If needed, `iter_chunks` can also be exposed individually via a streaming command `dvc stream` (or some other name).

Imagine the following command:
```
class CmdDataTransfer(CmdDataBase):
    def run(self):
        try:
            stats = self.repo.transfer(
                target=self.args.target,
                target_remote=self.args.target_remote,
                source=self.args.source,
                source_remote=self.args.target_remote,
                chunk_size=self.args.chunk_size,
            )
```

Being handled by `remote.base.Remote._process` as:
```
if transfer:
    # Threading, progress bar code omitted
    func = partial(self.tree.transfer, *args, **kwargs)
    executor.map(func, to_info, from_info, chunk_size)
```

Which calls `tree.transfer`, for example in `tree.webhdfs.WebHDFS`:
```
def transfer(self, to_info, from_info, chunk_size):
    read_stream = from_info.tree.iter_chunks(from_info, chunk_size)  # NOTE: Assuming a BaseTree subclass instance set as an attribute to simplify the example
    with self.hdfs_client.write(to_info.path) as writer:    
        for chunk in read_stream:
            writer.write(chunk)

def iter_chunks(self, from_info, chunk_size):
    return (chunk for chunk in self.hdfs_client.read(from_info.path, chunk_size=chunk_size))
```

The `tree.s3.S3` version needs a bit more work:
```
def transfer(self, to_info, from_info, chunk_size):
    import boto3

    read_stream = from_info.tree.iter_chunks(from_info, chunk_size)

    s3 = boto3.resource('s3')
    _id = self._get_multipart_upload_id()  # NOTE: Implementation needed. Generate a unique id for this upload
    multipart_upload = s3.MultipartUpload(to_info.bucket, to_info.path, _id)  
    
    for n, chunk in enumerate(read_stream):
        part = multipart_upload(str(n)) # NOTE: Restrictions apply: parts must be at least 5 MB, no more than 10k parts.
        part.upload(Body=chunk)  # NOTE: Error handling is missing, multipart_upload must always be closed to avoid charges.
    multipart_upload.complete()

def iter_chunks(self, from_info, chunk_size):
    return (chunk for chunk in self._get_obj(from_info)["Body"].iter_chunks(chunk_size))
```

What do people think? More importantly: is the problem properly identified? 

And to a lesser extent: does the solution outlined have any potential? I outlined how things would look for S3 and WebHDFS since those two are probably the ones I have most present now, but from a few quick Google searches similar implementations using [Azure SDK](https://github.com/Azure/azure-sdk-for-python/issues/74), `paramiko` (for SSH), and anything built on top of `requests` are possible.

**Comment by efiop** [comment]
Hi @tomasfarias ! Thanks for looking into this!

> Let me know if I understand the problem clearly: although datasets in external drives/s3/etc can be dvc added with --external and later dvc pushed to a remote, we would like to streamline the process via a new feature(s).

There are some exceptions to the `--external && push` relations, but you are in the right direction :ok_hand: The problem you are talking about is transferring to remote (currently `download + dvc add data + dvc push`), but there is another part about transferring to cache (currently `copy + dvc add data`). I see that you are already diving into implementation details, but let me ask you a few high-level questions first.

Could you please show how such scenarios would look like in terms of commands that you've suggested? I see the draft for CmdTransfer, but would be interesting to see it in a higher-level description. E.g. previously it was
```
aws s3 cp s3://bucket/data data --recursive
dvc add data
dvc push
```
for straight-to-remote and for straight-to-cache it was:
```
# dvc repo is on SSD, but the data is on hdd
cp /hdd/data data -a
dvc add data
# might want to use cache.dir on hdd and cache.type `symlink` to link from cache on hdd to ssd workspace
# https://dvc.org/doc/user-guide/large-dataset-optimization
```
but now if the `data` doesn't fit on my local machine (for the first scenario) or on my SSD (for the second scenario), the new workflow will be...?

**Comment by tomasfarias** [comment]
Hi @efiop! Thanks for the comments.

Exactly as you pointed out, I was mostly focusing on the first problem, i.e. transfer straight-to-remote, but haven't really thought too much about the second one, i.e. transfer straight-to-cache.

So, starting with the first problem, since I have it more present, in terms of commands I imaging things looking like:
```
dvc remote add hdfsremote hdfs://user@hdfs.com/data
dvc remote add s3cache s3://bucket/data
dvc transfer --from=s3cache --to=hdfsremote --recursive
```

This would transfer all files in the `s3cache` remote over to `some_remote` remote storage. Transfering a single file would eliminate the `--recursive` flag and require a file path instead:
```
dvc remote add hdfsremote hdfs://user@hdfs.com/data
dvc remote add s3cache s3://bucket/data
dvc transfer data.csv --from=s3cache --to=hdfsremote 
```

In this case, `hdfsremote` would be setup as the new external cache for the file(s) transferred, what `dvc add --external` would do.

I'm more than open to brainstorming and hearing opinions about more intuitive names for the command line arguments, for example:  `--source`/`--target` or `--cache`/`--remote`, this was just an initial idea.

Additional command-line options that are available in other commands should be supported in `transfer`, for example: `--glob` as used in `add`, and `--jobs` as used in `push`. This makes sense since `transfer` essentially combines the functionality of the two. 

For the second problem, an external cache should be setup in the HDD first, and passed as the `--from` parameter. `--to` can be ignored since we would only be replicating what `dvc add` does without the `dvc push` part:
```
dvc remote add hddcache /hdd/data
dvc transfer data.csv --from=hddcache 
```

I'm struggling a bit to think how would this be supported though. In the event that a `dvc push` command runs next, I'm thinking we need to find a way to signal the data should be moved in chunks as if `dvc transfer` with `--to` was ran right from the start. Potentially, some additional metadata could be stored in the `*.dir` file to signal which specific files should be transferred in chunks from a source external remote. Do you think this has any potential? 

Additionally, do you have any comments regarding the high level workflow? Is there something I may have not understand correctly?

**Comment by efiop** [comment]
@tomasfarias Thanks! Ok, then let's focus on `straight-to-remote` for now, for simplicity. And let's clarify some terminology first. I see that there is clear confusion between remote/cache and storage in your examples. dvc remote/cache are always in content-based format, like `.dvc/cache` is, so they are not human readable. I think the reason is that you were looking for a rather less known `remote://` notation that we have, that, amid the confusing name, is used as a shortcut for urls that might require some special config options. E.g. your example of:
```
dvc remote add hdfsremote hdfs://user@hdfs.com/data
dvc remote add s3cache s3://bucket/data
dvc transfer data.csv --from=s3cache --to=hdfsremote 
```
could be directly rewritten(not changing other aspects, just armed with remote:// tribal knowledge) like so:
```
dvc remote add hdfsremote hdfs://user@hdfs.com/data
dvc remote add mys3 s3://bucket/data
dvc transfer remote://mys3/data.csv --to=hdfsremote
```
but I also see that you're using `data.csv` as an argument for `dvc transfer` as an output name, right? So it would create `data.csv.dvc` in the current directory? What if I want to place it at some arbitrary path in my repo with some arbitrary name? Maybe existing commands like `dvc add/import-url` could be used to firgure out a more natural sematic for `dvc transfer`? WDYT?

**Comment by tomasfarias** [comment]
@efiop Thanks for clarifying! I was definitely a bit confused. 

Yeah! My whole intention was to have the argument `data.csv` indicate `data.csv.dvc` was to be created in the current directory. I do think that is limited, and actually after reading over a bit more about `dvc import-url` seems like it would fit the task exactly as I was thinking about and it already supports an arbitrary target path via the `out` argument. Maybe the same signature could be used for `dvc transfer`?

```
dvc transfer s3://bucket/data some/out/path/data
```

This would move the file in chunks over to some arbitrary path `some/out/path/data`. Going back to the first example, this path should be an HDD with plenty of space or some other external storage. Does this make more sense?

**Comment by efiop** [comment]
@tomasfarias That sounds good!

**Comment by isidentical** [comment]
Hey @efiop, is this copy/tranfer operation will be on a new command or an already existing one?

Also, may I request 2 workflow command-group samples (some sort of small, self-contained replication of the whole process so that I can test one my local environment)? One is from the current workflow of people, where they try to do on 2 different local file system points, and one is from what they are going to execute when this issue will be resolved.

**Comment by efiop** [comment]
@isidentical I'm not a 100% sure if this might fit into some existing command, would love to hear your thoughts on it.

Sure, here are some examples. Please make sure you are familiar with our workflow already (see our get-started guide)
### `straight-to-remote` scenario:
Imagine having some file/directory on s3(or other cloud) and you want to add it to your dvc repo and push to your remote (say it is an s3 remote too, say `dvc remote add -d mys3 s3://bucket/dvc-remote`), you would need to
```
aws s3 cp s3://bucket/path/to/data data
dvc add data
dvc push
```
but now what if `data` is so big that it doesn't fit as a whole on your machine? You won't be able to perform the same set of commands. DVC is able to pull partial data(e.g. if `data` is a diretory: `dvc pull data/subdir`) or provide you streaming access to the data on remote (e.g. if `data` is a file, you could use [api.open/read](https://dvc.org/doc/api-reference/open)) but we need to somehow shove that giant file into the remote first. So we need some way to put that data into remote without downloading it fully locally. Again have to stress that this new operation, when it is done, should leave the `dvc` repo in a state that is indistinguishable from that regular worklow described above with the exception that you won't have that data in the local cache or in your workspace, only present in the remote. Regarding particular command name and semantics - it is up to you to research and suggest the CLI design for it :wink: 

### `straight-to-cache` scenario:
Now imagine you have giant HDD and a small SSD that you keep your dvc repo(aka workspace) on. Since SSD is too small, you set up a dvc cache on HDD (see https://dvc.org/doc/use-cases/shared-development-server) (say /storage/dvc-cache) and use symlinks to link from cache into your local workspace (see https://dvc.org/doc/user-guide/large-dataset-optimization). Now say you have a giant dataset in /storage/some/path/data that you want to add to your repo as if you'd be able to:
```
cp /storage/some/path/data data
dvc add data
```
your first instinct might be to `dvc add /storage/some/path/data` but that file is outside of dvc repo, but dvc will complain. At this point people often misuse `--external` and `dvc add /storage/some/path/data --extranal`, but that will make `/storage/some/path/data` a part of your workspace (dvc will change it there) instead of having `data` in your dvc repo. So it would be nice to have a way to transfer that file to your workspace. The solution for this feature might be related to straight-to-remote solution, or it might be a separate command/flag, it is unclear right now. If you wish, we can focus on one straight-to-remote or straight-to-cache exclusively for now, to narrow the scope, but it is just useful to keep in mind that these might have a common solution/interface.

**Comment by isidentical** [comment]
As a command-line model, I am thought something like this;
```
dvc transfer [--to-remote / --to-cache] <file>
```

```
# straigh-to-remote's example
dvc transfer --to-remote s3://bucket/path/to/data data
# straight-to-cache's example
dvc transfer --to-cache /storage/some/path/data
```

**Comment by isidentical** [comment]
I am a bit blocked on the user interface, even though I can start writing tests and migrate them later I'd rather prefer to work on the interface first. Here are the ideas that came and go during meetings
- `dvc export-to-remote <url> <name>? -r <remote>`
- `dvc add <name> --straight-to-remote --from-url <url> -r <remote>`
- `dvc import-url <url> <name> -straight-to-remote -r <remote>`

I see amending to the `dvc add` as the least likely option, since it would be too complicated to adjust it with 2 mandatory flags that one doesn't work with the other one (`--straight-to-remote`, `--from-url`). 

@efiop said `import-url` tracks the original location where we would ignore it when straight-to-remote functionality is activated, so it might also not be the ideal one. 

For `export-to-remote`, even though it is a new command @shcheklein suggested semantics are not clear from the name which I definitely agree with. Maybe we could re-use the `import` keyword, something like `import-to-remote` (since as @efiop already stated, this is what it does) and also `import` already associated with both `track` + `download` operations.

**Comment by efiop** [comment]
Straight-to-remote would be useful both in `dvc add` scenario and in `dvc import-url` actually, and in the latter one it should still track the original location as it does right now. This actually makes me wonder if it should indeed be a flag for both `dvc add` and `dvc import-url`. E.g.
```
dvc add s3://bucket/path -o mydata --straight-to-remote -r myremote  # would create mydata.dvc or path.dvc if no -o specified
dvc import-url s3://bucket/path mydata --straight-to-remote -r myremote  # would create mydata.dvc or path.dvc if no out is provided
```
the name of the flag looks a bit ugly though :slightly_smiling_face: And i'm not sure how intuitive it will be for users, though those same users find `--external` somehow and misuse it all the time :slightly_smiling_face: 

Notice how I didn't use `--from-url` in `dvc add`, because it seems unintuitive and users already misuse `--external` with `dvc add s3://bucket/path --external`, so that might be the intuitive way to add something from external location.

Regarding `dvc export-to-remote`, it creates an association with uploading dvc-tracked data to an external location, at least for me. Kinda like if you need to deploy a model to s3 from your dvc repo.

The cons of `--straight-to-remote` for `add/import-url` is that they will be overloading the add/import-url CLI, which might be too much. There might be a better way, but if not - this one seems acceptable, even if just in experimental mode, until we find a better name.

**Comment by efiop** [comment]
For the record: while discussing with @isidentical , we found that in straight-to-cache scenario something like
```
dvc add /external/file -o data
```
would do what user expects: it will cache the `file`(aka transfer it to cache) and will make it known as `data` in the root of the repo. If the user has configured external cache with, say, symlinks, then `data` will be a symlink too, so you can add a giant file to your project this way.

There is a weird association with `--no-pull` for `dvc add/import-url` when discussing `--straight-to-remote`, but that's not quite it. Maybe `--transfer`? Not finding a perfect name/ui here so far :slightly_frowning_face:

**Comment by isidentical** [comment]
> There is a weird association with --no-pull for dvc add/import-url when discussing --straight-to-remote, but that's not quite it. Maybe --transfer? Not finding the perfect name/ui for it so far slightly_frowning_face

After a bit more discussion, we decided that just bare `--to-remote` (without `straight`) also might be an option. The problem with it is, the combination of it with `-r` looks a bit odd `dvc add s3://bucket/file --to-remote -r my_other_remote`. We could potentially reduce both `--to-remote` and `--remote` into just `--to-remote <remote (optional)>` but it wouldn't be much consistent with actions (pull, push etc).

**Comment by jorgeorpinel** [comment]
Hi! Re-discovered this from docs (and mentioned in demo). Some comments and questions:

> but also for dvc import[-url]
> import-url tracks the original location where we would ignore it when straight-to-remote functionality is activated ...something like `import-to-remote`
> it should still track the original location

@efiop Wouldn't #4527 be a requirement for applying this to imports?
Otherwise indeed, it doesn't seem to make sense to enable this for `import[-url]` (if the result is exactly the same as for `dvc add`).

Also, the `import-to-remote` only sounds correct for the import case, what if you don't want to save the original location (just-add case)? Cc @isidentical 

Personally I would focus on `add` only at first, but up to you.

> straight-to-remote scenario... `aws s3 cp s3://bucket/path/to/data data; dvc add data; dvc push`

Again, wouldn't #4527 cover that already? Maybe you could even combine `import --no-exec` with `--store` to reach the same "straight-to-remote" outcome.

> straight-to-cache scenario

This is only meaningful when you have a custom cache outside the DVC project, right? Does that have any special implementation implications? (Prob not.)

> `dvc transfer`
> `dvc export-to-remote`

@isidentical those names sound more like a disconnected utility, similar to `dvc get` (cc @shcheklein). If this is about tracking data in the process (just skipping placing it in the cache and/or workspace) then I agree (as implemented) that it makes more sense as a flag for `dvc add`.

> `add --straight-to-remote --from-url ...` ... too complicated to adjust it with 2 mandatory flags

Agree on avoiding `--from-url`  — by implying `--external` in `--to-remote` if I understand this correctly.

> users find --external somehow and misuse it all the time

DVC gives a hint to try that, I think 🙂 BTW would we be able to give a hint about `--to-remote`?

> `dvc add /external/file -o data` would do what user expects

Except that you can't check it out to the workspace, right? But def. there are several similar operations and maybe we should reconsider them to see how they can be consolidated... Probably won't be obvious which one is appropriate to most users.

> Maybe --no-pull or --transfer? Not finding the perfect name/ui for it so far

I think just `add --to-remote` is short and accurate enough. Another idea is to have a flag for each case: `--no-cache` or `--skip-cache` [straight to remote] AND `--no/skip-local/workspace/wdir` [straight to cache] (the former includes the latter).

> We could potentially reduce both --to-remote and --remote

I definitely agree on letting --to-remote take an optional remote name as argument instead.

**Comment by jorgeorpinel** [comment]
So I answered some of my questions (~~scratched~~ above) reviewing UI in #5198. I have one thing to add though, after realizing that `add --to-remote` includes `push`.

1. Is this too much functionality in `add`? Maybe a standalone command for this is best after all... (`dvc store/backup [--external]`?)
2. What would be the result if pushing fails (does `dvc push` complete the operation after fixing the remote or do you have to try again with `-f`) ?
3. Is there any alternative workflow e.g. `import --no-commit` first, and push separately (again, maybe via #4527) ? I think it would be ideal to understand all the ways we deal with external data and try to consolidate them because it's getting confusing 🙂

**Comment by isidentical** [comment]
I tried to answer a few questions, though @efiop probably has more context than I regarding this

> @isidentical Personally, I would focus on add only at first, but up to you.

I would be fine with either only one (add/import-url) or both of them. Though I feel like `import-url` is actually better fit to task, since it already covers tracking and downloading data. The only different semantic we would be introducing is, instead of downloading it to the workspace, it will download it to the remote storage. 

> Notice how I didn't use --from-url in dvc add, because it seems unintuitive and users already misuse --external with dvc add s3://bucket/path --external, so that might be the intuitive way to add something from external location.

Though as @efiop already stated, one reason that we are adding this is to solve the confusion 

> @isidentical those names sound more like a disconnected utility

Definietly, and it is why we gave up on those.

> I definitely agree on letting --to-remote take an optional remote name as argument instead.

It wasn't much consistent with other commands that worked with remote storage (push/pull etc) so we ended up with having `-r`,`--remote` as a  separate option.

> What would be the result if pushing fails

What would you mean by result, like in which end? If it is related the dvc file creation, no we don't create files before we complete the transfer operation.

**Comment by shcheklein** [comment]
@jorgeorpinel 

> `add --to-remote includes push`

it doesn't include push, it looks like it from a user perspective, but it's not `dvc push` . In this case it's still add, but instead of caching it saves data to remote directly (and skips cache). 

> does dvc push complete the operation

it's not dvc push, so hope it answers it

> Is there any alternative workflow e.g. import --no-commit

can be, but in this case you will have to do something like `dvc push --from-remote path ` or something like this. Or a separate command `dvc transfer` that would do only data transfer.

I guess, the logic behind using `dvc add` was to try to give user the most expected interface `dvc add s3://something` - it should explain them two options - to-remote and external.

----------------

> Though I feel like import-url is actually better fit to task, since it already covers tracking and downloading data.

@isidentical @efiop  hmm, in this case I still don't quite understand why did you decide to touch `import-url`. It has a very different semantics (which you can see in the DVC file it creates). `import-url` doesn't put anything into DVC control. It just track changes in some external URL. It doesn't cache, it doesn't save into remote, etc. So, I'm still trying to understand the semantics of `dvc import-url s3://something --to-remote`. What is the use case here? @efiop could you clarify what were you thoughts behind it? And can we start with `dvc add` first for simplicity?

**Comment by jorgeorpinel** [comment]
Thanks for all the answers! It's def. becoming clearer. One last round from me:

> one reason that we are adding this is to solve the (--external) confusion

@isidentical sure, and that's a good idea. Though I think this adds yet another option to dealing with external data so not sure the confusion will improve. Options now: add --external vs add --to-remote vs import --no-exec vs... — may be worth re-thinking how they overlap and consider redesigning — prob. out of scope for now.

> It wasn't much consistent with other commands that worked with remote storage (push/pull etc) so we ended up with having -r,--remote as a separate option.

Using the exact flag names of different commands doesn't necessarily mean consistent. And that principle probably shouldn't trump usability. I find `add --to-remote --remote` repetitive and `add --remote` by itself is meaningless (forbidden in fact). But what about simply `add --remote <name>` (best of both worlds)?

> it's not dvc push ... instead of caching it saves data to remote directly

@shcheklein I see what you mean but it still feels like it's really push (maybe it's even used internally?), because add never stored data to remote storage before, so "skips cache" still doesn't quite explain it. `add --no-commit" is what I associate with `add` skipping cache.

And I see that `push` doesn't complete a failed `add --to-remote` (because no .dvc file is created, which is good). But what if you clone a repo where this was done, and use `push` (or `pull`)? Does it try to transfer from the original location to the remote (again)? _i.e. How does sync ops interact with to-remote .dvc files?_

> Or a separate command dvc transfer that would do only data transfer

Yeah I'm starting to incline for this TBH (suggested names `dvc store/backup` which align with #4527) but I guess there's no perfect solution except perhaps redesign the whole external data management layer (which we can still do in the future).

> still don't quite understand why did you decide to touch import-url

I think I understand why this can be applied to `import-url` (again, assuming we're throwing in #4527 internally), but I would ask why not also `import`? Meaning: to move specific data from one remote to another, while at the same time tracking it independently in a new project (an alternative to [chained imports](https://github.com/iterative/dvc/issues/3305)).

**Comment by efiop** [comment]
Thanks for great questions guys!

Indeed `import-url` was a logical continuation, which led us to better understand the feature in general and how it could be used in both add and import-url, which lead us to a conclusion that a new `transfer` command won't cut it, as it will then have to duplicate logic and semantics of both add and import-url. To simplify the first iteration, we could indeed drop `import-url` for now, only focusing on `add`. As noted by @jorgeorpinel `dvc import` doesn't support backups yet, so `--to-remote` doesn't make sense for it for now.

> @isidentical @efiop hmm, in this case I still don't quite understand why did you decide to touch import-url. It has a very different semantics (which you can see in the DVC file it creates). import-url doesn't put anything into DVC control. It just track changes in some external URL. It doesn't cache, it doesn't save into remote, etc. So, I'm still trying to understand the semantics of dvc import-url s3://something --to-remote. What is the use case here? @efiop could you clarify what were you thoughts behind it? And can we start with dvc add first for simplicity?

@shcheklein `import-url` does put the downloaded file under DVC control. And `dvc import-url`(unlike `dvc import`) does cache and does push to remote. So it makes sense to have --to-remote there too.

**Comment by shcheklein** [comment]
> @shcheklein import-url does put the downloaded file under DVC control. And dvc import-url(unlike dvc import) does cache and does push to remote. So it makes sense to have --to-remote there too.

Yep, makes sense. Got confused by a bug in our docs (and this constant struggle to remember import-url vs import):

`⚠️ DVC won't push or pull imported data to/from remote storage, it will rely on it's original source.` cc @jorgeorpinel 

https://dvc.org/doc/command-reference/import-url

I would mention this semantics (is data put into cache + saved into remote somewhere at the top of that page)

Up to you guys then, to do import-url or not.

**Comment by jorgeorpinel** [comment]
> dvc import doesn't support backups yet, so --to-remote doesn't make sense for it for now

Ah I see. I didn't realize that `import-url` **does** support pushing to remote by default! OK, we got that wrong in docs, will fix (and move the note up for the `import` ref, @shcheklein).

**Comment by jorgeorpinel** [comment]
To recap, my main remaining concern is with the (inexact) idea that `--to-remote` seems to "include push" — maybe not push, but a new form of sync (call it `transfer`). It feels that way to me because add/import never stored data to remote storage before. And "Skipping caching" I associate with `--no-commit`. So

1. Can you please confirm that the implementation doesn't download anything internally, @isidentical? (I assume so since the use case is that you may not have enough local storage space, just double checking 🙂).

2. What if I `add --to-remote` something and then accidentally delete the copy from remote storage; Is there a way to re-transfer the data other than deleting the .dvc file and adding again?

3. Similar Q for `import-url --to-remote`: will `update` "skip the local" (transfer to remote storage)? i.e. is it a new field in the .dvc file dependency? Or will that actually download to the local project.

4. Finally, to get the data locally after either command, what do you do; just `dvc pull` it?

Thanks

**Comment by isidentical** [comment]
> Can you please confirm that the implementation doesn't download anything internally, @isidentical?

It does, at least partially. Either it downloads a chunk of the file, and uploads it (if the provider supports it, only s3 + ssh + local does as of right now) or it does the whole file and uploads it (not whole dataset, but whole file, depending on the thread count the user system can store 8-10 files at the same time if they are uploading a remote where chunked upload isn't supported). 

> What if I add --to-remote something and then accidentally delete the copy from remote storage; Is there a way to re-transfer the data other than deleting the .dvc file and adding again?

We don't check whether cache exists on the remote or not (since for checking that, we need to actually have the all hash_infos, which means we have to download it again, so we directly transfer whether it already exists or not)

> Similar Q for import-url --to-remote: will update "skip the local" (transfer to remote storage)? i.e. is it a new field in the .dvc file? Or will that actually download to the local project (and you need to dvc push separately)?

Good point, we actually don't have any special case regarding `update` so it behaves like imported something and want to sync and it makes a local clone (which I don't think the desired behavior, it is open for recommendations)

> Finally, to get the data locally after either command, what do you do; just dvc pull it?

Yes

**Comment by jorgeorpinel** [comment]
> or it does the whole file and uploads it... user system can store 8-10 files

Won't this defeat the purpose of `--to-remote` in cases where the dataset in question is <= 10 files and the remote doesn't support chunking?

I don't oppose the feature over this at all, but I'm wondering if there should be a follow-up ticket to address that, e.g. to leverage built-in chunking once #829 is solved, perhaps. And meanwhile maybe disable this for the cases where it doesn't make sense?

> We don't check whether cache exists on the remote

Ok so if I try `pull` and no data is found, the .dvc file is useless, right? I wonder what's the best way to communicate this to the user (that they need to delete it and use `add --to-remote` again.

> to actually have the all hash_infos

This is an interesting point too. So the .dvc file won't have any hash values after `add/import --to-transfer`, that means they'll be considered `modified` immediately by `status`. Or should there be a new state introduced for this?

> we actually don't have any special case regarding update... it is open for recommendations

Maybe (esp. if we intro a new `status`) `update` can fail with an error on to-remote imports unless they have been pulled before.

**Comment by isidentical** [comment]
> Won't this defeat the purpose of --to-remote in cases where the dataset in question is <= 10 files and the remote doesn't support chunking?

It seems like all providers support chunking, and the progress on its way to being implemented. So no need to worry about this anymore! 

Regarding `update`/`status` behavior, I'll investigate and try to come up with possible solutions!

**Comment by isidentical** [comment]
> This is interesting too. So if to-remote .dvc files won't have hash values at first, that means they'll be considered modified initially (by status).
```
$ dvc add remote://default/data --to-remote -r default
100% Add|██████████████████████████████████████████████████████████|1/1 [00:00,  3.31file/s]
                                                                                            
To track the changes with git, run:                                                         
                                                                                            
        git add data.dvc                                                                    
$ dvc status
data.dvc:                                                             
        changed outs:
                not in cache:       data
$ dvc pull -r default
A       data/                                                                               
1 file added and 1 file fetched       
$ dvc status         
Data and pipelines are up to date.                                    
```

The current behavior says they are `not in cache`, which is actually true. I guess only thing we can do is somehow make it explicit? like `not in local cache`? Or just keep it as is

For `dvc update`, I believe implementing the same flag (`--to-remote`) should be sufficient. Though I agree that without that flag, the behavior would be still a bit unorthodox. IIRC @efiop 's initial idea about this flag was because it is the simplest solution without introducing a new field to the dvc files.

**Comment by jorgeorpinel** [comment]
Hey sorry for the delay.

> The current behavior says they are not in cache, which is actually true. I guess only thing we can do is somehow make it explicit? like not in local cache?

I like the idea of a new status/message for this, yes. I was thinking more about the `changed outs:` grouping, as they're not "changed". Maybe `transferred:`? Idk if it complicates things too much.

**Comment by efiop** [comment]
Support for remote status in regular `dvc status` is something that we've been talking about for a while https://github.com/iterative/dvc/issues/5369 . It will naturally support transfered data once we implement it. Def not part of this issue.

**Comment by efiop** [comment]
@isidentical https://github.com/iterative/dvc/pull/5343 is the last part we are missing (not counting for docs) until we can close this ticket, right? Just checking.

**Comment by isidentical** [comment]
> @isidentical #5343 is the last part we are missing (not counting for docs) until we can close this ticket, right? Just checking.

Yes

**Comment by jorgeorpinel** [comment]
> Support for remote status in regular dvc status is something that we've been talking about... Def not part of this issue.

Sure. I'm not suggesting that. I'm saying that the "change outs" message is misleading. No outputs have changed. But I guess that can happen in other circumstances so also out of scope here. But do you think we should try to address this (separately)?

**Comment by efiop** [comment]
@jorgeorpinel Sorry, didn't mean to dismiss it like that :slightly_frowning_face: That was just for the record, no bad intention.

Sure, we should reconsider the status output, it is quite obsolete. Doesn't seem like there are good simple action points there though, we'll need a general redesign of output to address multiple complaints that we've collected over the years. Looks like it is better to create an epic to collect all complaints instead of creating a yet another ticket. Though we can always repurpose it for an epic, no problem.

**Comment by jorgeorpinel** [comment]
Sounds good!

**Comment by jorgeorpinel** [comment]
Howdy again! Something came up in https://discord.com/channels/485586884165107732/563406153334128681/806673366416883733 about this recently: what happens if you setup a remote, then set is as external cache e.g. with `dvc config cache.local` (or any other kind), and finally `add/import` something straight `--to-remote`? I should try before asking but don't have time this instant. Seems like an unintended confusion or maybe even a possibly problematic circular situation.

**Comment by efiop** [comment]
@jorgeorpinel Nothing, it all gets transfered to remote directly.

**Comment by jorgeorpinel** [comment]
OK its just confusing because the result is that the data gets "transferred" to the remote location, which is your (external) cache, so the data gets "cached" but without calculating hashes. I wonder if `dvc pull` skips the non-download and then no hashes are ever calculated, or something like that.

**Comment by jorgeorpinel** [comment]
I guess you can just use `dvc checkout`? Sorry I'm just thinking out loud. Should try this... The point is that there may be unintended side effects to keep in mind.

**Comment by efiop** [comment]
@jorgeorpinel `dvc pull`, you need to download it from the remote first.

**Comment by alealv** [comment]
Hi folks,

I was the one discussing with @jorgeorpinel  on the discord channel. I used DVC in the past so I have some experience with it.
At this new project, I face the mentioned `straight-to-cache` issue.

We use a server to train ML models with many GPUS, hence I configured DVC as [Shared Development Server](https://dvc.org/doc/use-cases/shared-development-server). The project/workspace is on a small SDD and the DVC cache is on a big NAS HDD. 

The problem is that the pipeline I created to process the raw dataset (~3TB) and generate the data use to train (~6TB) is too much for the SDD, and I need/want to have it inside the project folder. The solution is to have a symlink pointing to the data in the NAS.

### Tracking data directly

Before using the DVC pipeline I tracked the data with a `data.dvc` file and achieved what I intended in the following way:

```sh
dvc config cache.type hardlink,symlink
dvc add /mnt/nas/data
dvc move /mnt/nas/data ~/project/data

```
I guess the `straight-to-cache` solution you were proposing would simplify the process

### Using a pipeline

But now, I created a pipeline which downloads (raw data comes from public datasets) and process the data. In this scenario I have a `dvc.yaml` file similar to this one (I don't put the original because it has 158 lines):

```yaml
stages:
  download-data:
    cmd: wget https://<some-repo> -P /mnt/nas/raw_data
    outs:
      - /mnt/nas/raw_data:
      	cache: false
        persist: true

  proc-data:
    cmd: ./proc_data.sh --input /mnt/nas/raw_data --output /mnt/nas/waves
    outs:
      - /mnt/nas/waves
```

This works fine, but it's tracking the `/mnt/nas/waves` directory, which is outside the project.

If I do the following:

```yaml
...
  proc-data:
    cmd: ./proc_data.sh --input /mnt/nas/raw_data --output waves
    outs:
      - waves
```

It will track the data where I intended, having a symlink `waves -> /mnt/nas/dvc-cache/0f/<hash>.dir`, which is what I want. But not before generating all the wave files in the SDD and then transferring them to the external cache.

That's why the `straight-to-cache` option will be optimal. But as I understand you are considering this for some commands but not for the pipeline, I'm I wrong?

### Having a `local remote` as `cache`

I also asked @jorgeorpinel about this thread and configuring the cache as a remote, because the documentation says:

_cache.local - name of a local remote to use as a custom cache directory. (Refer to dvc remote for more information on "local remotes".) This will overwrite the value provided to dvc config cache.dir or dvc cache dir._
https://dvc.org/doc/command-reference/config

I guess with that and this feature it will behave as I wanted.

To be honest, this confuses me a little, about the difference between a `local remote` and the `cache`. AFAIK, if I have a `local remote` and a `cache` on the same disk but different folders, I would have a double copy of the data, Am I right? I have not considered _merging_ these two, and what are the implications of that.

Sorry for the long explanation, but I wanted to be very clear.

**Comment by jorgeorpinel** [comment]
Hey Ale nice chatting with you and thanks for posting here instead for visibility. My comments:

> Before using the DVC pipeline
dvc add /mnt/nas/data
dvc move /mnt/nas/data ~/project/data

I think you meant `dvc add --external` there. BTW that's an interesting use of `add` + `move`! 👍 

> I guess the straight-to-cache solution you were proposing would simplify the process

I think that `import-url` by itself (no need for `--to-cache`) already achieves the same result: copies the data to the (external) cache first, then links it locally (assuming file links are configured in DVC and supported by the FS).

Or, if you meant setting up a remote with the same path as your external cache (error-prone), and then using `add --to-remote`. As you can see you still need 2+ steps, and things get a little confusing. So not sure... It's def. another interesting command interaction! But probably not the intended use here (hacky).

> That's why the straight-to-cache option will be optimal. But as I understand you are considering this for some commands but not for the pipeline, I'm I wrong?

Correct, this issue doesn't affect pipelines at all. But I guess it's a good idea to consider implementing `to-cache/remote` for dvc.yaml `outs`. Thoughts @efiop @isidentical ?

<details>

<summary>Expand to see the rest of the discussion</summary>

> The rest isn't related to this issue so I hid it.

> now, I created a pipeline
tracking the /mnt/nas/waves directory, which is outside the project  brings some problems...

1. Actually, there should be no problem for users to `checkout` different versions from the same cache. That's what the share cache patters is for 🙂 (The only issue could be file system errors if users try to write the same file at once — unlikely)
2. Is there absolutely no way for your scripts to read from `/mnt/nas/waves` instead? Otherwise, my recs are to add a stage, or an extra command in the training stage (`cmd:` can [contain a list](https://github.com/iterative/dvc/issues/4373)) that creates the link for now.

> `proc_data.sh --input /mnt/nas/raw_data --output waves` will track the data where I intended  But not before generating all the wave files in the SDD

Yes, DVC doesn't change your code and can't capture/redirect file system write operations, at least for now (we accept feature requests).

> the documentation says:
*cache.local - name of a local remote*... this confuses me a little

Yeah that doc needs an update, thanks for the heads-up! I'm starting to review it in iterative/dvc.org/pull/2154

> if I have a local remote and a cache on the same disk but different folders, I would have a double copy of the data, Am I right?

You are correct.

> merging these two, and what are the implications of that

Not a good idea 🙂 — the cache could potentially get corrupted, and it's just confusing! We do sin a little in that we kind of do just that (merge the remote and cache concepts) for [external outputs](https://dvc.org/doc/user-guide/managing-external-data#examples), which is part of the reason why we don't really recommend that anymore.

</details>

**Comment by alealv** [comment]
> think you meant `dvc add --external` there. BTW that's an interesting use of `add` + `move`!

Yes, my bad. I forgot the `--external`

> Actually, there should be no problem for users to checkout different versions from the same cache. That's what the share cache patters is for slightly_smiling_face (The only issue could be file system errors if users try to write the same file at once — unlikely)

I think there will be because when checking out the folder being tracked will change (aka `/mnt/nas/data`) and this is shared through all users.

Thanks for solving my other doubts.

I could modify the training script but it's not a good idea. What I will do is manually symlink the data folder inside the project.

**Comment by jorgeorpinel** [comment]
> when checking out the folder being tracked will change (aka /mnt/nas/data

`dvc checkout` doesn't alter the cache at all. It reads .dvc files and dvc.yaml/lock files and creates the appropriate links FROM the cache into your workspace 🙂

**Comment by alealv** [comment]
> dvc checkout doesn't alter the cache at all. It reads .dvc files and dvc.yaml/lock files and creates the appropriate links FROM the cache into your workspace slightly_smiling_face

Exactly, I wasn't referring to the cache. `/mnt/nas/data` will be changed and this is shared across all users. If I'm training and somebody checks out to a different commit it will corrupt my training.

**Comment by jorgeorpinel** [comment]
Ah I see what you mean. Yes, the external data itself should not be shared by several projects! External outputs are considered part of the extended workspace and project workspaces can't overlap, naturally. That's another reason why we don't recommend external outputs except for very specific uses where absolutely no other option is available.

> I guess https://dvc.org/doc/user-guide/managing-external-data shouldn't link to https://dvc.org/doc/use-cases/shared-development-server#configure-the-external-shared-cache for instructions to config an external cache, as **shared** external caches are not compatible with external outputs. Moved this to https://github.com/iterative/dvc.org/issues/654#issuecomment-775489955.

**Comment by shcheklein** [comment]
@alealv there is a workaround for this problem - https://github.com/PeterFogh/dvc_dask_use_case . You could use `remote://something` notation, and for different folks ask them to have their own `--local` , `--global`, or `--system` config to that remote that defines the exact location personally for them.
